core/src/androidMain/kotlin/Peripheral.kt

```diff
@@ -193,22 +193,22 @@ public class AndroidPeripheral internal constructor(
 
     /** Creates a connect [Job] that completes when connection is established, or failure occurs. */
     private fun connectAsync() = scope.async(start = LAZY) {
-        val connection = establishConnection().also { _connection = it }
+        try {
+            val connection = establishConnection().also { _connection = it }
 
-        connection
-            .characteristicChanges
-            .onEach(observers.characteristicChanges::emit)
-            .launchIn(scope, start = UNDISPATCHED)
+            connection
+                .characteristicChanges
+                .onEach(observers.characteristicChanges::emit)
+                .launchIn(scope, start = UNDISPATCHED)
 
-        try {
             suspendUntilOrThrow<State.Connecting.Services>()
             discoverServices()
             onServicesDiscovered(ServicesDiscoveredPeripheral(this@AndroidPeripheral))
             _state.value = State.Connecting.Observes
             logger.verbose { message = "Configuring characteristic observations" }
             observers.rewire()
         } catch (t: Throwable) {
-            closeConnection()
+            disconnect()
```

Comment:
I'm probably missing something, but what is gained with this change?
If we throw in `establishConnection` then `_connection` will still be `null` and `disconnect()` don't do anything more than `closeConnection()` did?

Reply:
This is unrelated to the first change.

My thought (and I could be wrong here) is that the callback system could cause a brief delay between calling `bluetoothGatt.disconnect` and actually updating to the `State.Disconnected`. Using `disconnect` instead of `closeConnection` will suspend until we have updated to that state, so we can't return from `connect` while still in the `Connecting` state.

If my initial assumption is wrong I'd be happy to remove this change.

Reply:
Another option is that I could keep this as `closeConnection` and instead we could add something like `if (_state.value !is State.Disconnected) { _state.value = State.Disconnected() }` to that function. That would have the added effect of defending us in the case that the callback is never invoked.

Reply:
> My thought (and I could be wrong here) is that the callback system could cause a brief delay between calling `bluetoothGatt.disconnect` and actually updating to the `State.Disconnected`.

Ohh, ya, I could definitely imagine a delay occurring there.

> Using `disconnect` instead of `closeConnection` will suspend until we have updated to that state, so we can't return from `connect` while still in the `Connecting` state.

Gotchya, now I'm starting to follow what the goal is.

> Another option is that I could keep this as `closeConnection` and instead we could add something like `if (_state.value !is State.Disconnected) { _state.value = State.Disconnected() }` to that function.

Is the `if` check necessary, would there be harm in always just setting `_state.value = State.Disconnected` in the `closeConnection()` call?

Reply:
Looking at it more closely, I think we do always want to set `_state.value = State.Disconnected` if an exception occurs.
Otherwise if connection is rejected then we'd get stuck in `Connecting` state.

https://github.com/JuulLabs/kable/blob/6b4ca38fd30842c75fdf261e6f60602df2202501/core/src/androidMain/kotlin/Peripheral.kt#L191

...so, the combination of moving the `try` up and always setting to `Disconnected` on exception should protect against that.

With `_state` being `StateFlow` there isn't any harm in setting the same state multiple times.

Reply:
I have the `if` condition in there because `State.Disconnected` has a nullable `status` field I don't want to clobber.

Reply:
Oh, good point/catch.

Reply:
Updated to use `MutableStateFlow.update` to avoid race conditions.